src/coreComponents/physicsSolvers/fluidFlow/wells/SinglePhaseWellKernels.hpp

```diff
@@ -561,29 +569,21 @@ struct PresInitializationKernel
     real64 const avgDensity = MpiWrapper::sum( sumDensity.get() ) / numPerforations;
 
     real64 pressureControl = 0.0;
-    real64 gravCoefControl = 0.0;
-    if( isLocallyOwned )
+    real64 const gravCoefControl = refWellElemGravCoef;
+    // initialize the reference pressure
+    if( currentControl == WellControls::Control::BHP )
     {
-      // initialize the reference pressure
-      if( currentControl == WellControls::Control::BHP )
-      {
-        // if pressure constraint, set the ref pressure at the constraint
-        pressureControl = targetBHP;
-      }
-      else // rate control
-      {
-        // if rate constraint, set the ref pressure slightly
-        // above/below the target pressure depending on well type
-        pressureControl = ( wellType == WellControls::Type::PRODUCER )
-                        ? 0.5 * pres
-                        : 2.0 * pres;
-      }
-      gravCoefControl = wellElemGravCoef[iwelemControl];
-      wellElemPressure[iwelemControl] = pressureControl;
+      // if pressure constraint, set the ref pressure at the constraint
+      pressureControl = targetBHP;
+    }
+    else // rate control
+    {
+      // if rate constraint, set the ref pressure slightly
+      // above/below the target pressure depending on well type
+      pressureControl = ( wellType == WellControls::Type::PRODUCER )
+                      ? 0.5 * pres
+                      : 2.0 * pres;
     }
```

Comment:
Yes you are right rate-constrained wells have a `targetBHP` as well, and the code will stop if the `targetBHP` is not specified (for both single-phase and compositional). What is done here is just a not-very-good way to initialize the pressure in rate-constrained wells at the beginning of the simulation, before the first iteration. 

A better way would be to set the initial pressure using `targetBHP` (for instance, slightly below `targetBHP` for injectors). To be honest, I did not use the `targetBHP` to initialize the injectors that because I sometimes (often ;) ) use a very large `targetBHP` to keep the wells at rate constraint during the full simulation, even if it is not realistic (and I don't want this large `targetBHP` to perturb the initialization...). I am going to add this explanation to the inline comment just before the line you referenced. 

Ultimately, to have a really good well initialization and to improve the nonlinear convergence, it would be helpful to guess the well pressure and rate by solving small local well systems (with fixed reservoir values), but there is a little bit of work to do to allow that.